thermal-control-design.md

```diff
@@ -0,0 +1,272 @@
+    
+### 2.2 Fan device monitoring 
+
+In most case fan is the device to cool down the switch when the temperature is rising. Thus to make sure fan running at a proper speed is the key for thermal control.
+
+Fan target speed and speed tolerance was defined, by examining them we can know whether the fan reached at the desired speed.
+
+same as the temperature info, a [table for fan](https://github.com/Azure/SONiC/blob/master/doc/pmon/pmon-enhancement-design.md#153-fan-table) also defined as below:
+
+	; Defines information for a fan
+	key                     = FAN_INFO|fan_name              ; information for the fan
+	; field                 = value
+	presence                = BOOLEAN                        ; presence of the fan
+	model                   = STRING                         ; model name of the fan
+	serial                  = STRING                         ; serial number of the fan
+	status                  = BOOLEAN                        ; status of the fan
+	change_event            = STRING                         ; change event of the fan
+	direction               = STRING                         ; direction of the fan 
+	speed                   = INT                            ; fan speed
+	speed_tolerance         = INT                            ; fan speed tolerance
+	speed_target            = INT                            ; fan target speed
+	led_status              = STRING                         ; fan led status
+	timestamp               = STRING                         ; timestamp for the fan info fetched
+
+### 2.3 Syslog for thermal control
+
+If there was warning raised or warning cleared, log shall be generated:
+
+	High temperature warning: PSU 1 current temperature 85C, high threshold 80C！
+	High temperature warning cleared, PSU1 temperature restore to 75C, high threshold 80C
+
+If fan broken or become up present, log shall be generated：
+
+	Fan removed warning: Fan 1 was removed from the system, potential overheat hazard!
+	Fan removed warning cleared: Fan 1 was inserted.
+
+## 3. Thermal control management
+
+Adjust cooling device according to the current temperature can be very vendor specific and some vendors already have their own implementation. In below Appendix chapter describes a Mellanox implementation. But handle the cooling device according to some predefined policies can be generic, this is part of what Thermal control management will do.
+
+This cooling device control function can be disabled if the vendor have their own implementation in the kernel or somewhere else.
+
+### 3.1 Thermal control management flow
+
+It will be a routing function to check whether the policies was hit an the fan speed need to adjust, and also run vendor specific thermal control algorithm.
+
+
+Below policies are examples that can be applied:
+
+- Set PWM to full speed if one of PS units is not present 
+
+- Set PWM to full speed if one of FAN drawers is not present or one of tachometers is broken present 
```

Comment:
I guess this is the case where the FAN is apparently stuck at a non 0 low tacho reading (minimal threshold below which it is deemed faulty). You may want to add it to the policy configuration example.

Reply:
these are only some examples, the framework will allow the vendors to add their own specific policies.

Reply:
ok thanks